static/js/recent_senders.js

```diff
@@ -30,6 +32,61 @@ exports.process_message_for_senders = function (message) {
     stream_senders.set(stream_id, sender_message_ids);
 };
 
+exports.process_topic_edit = function (stream_id, old_topic, new_topic, new_stream_id) {
+    // When topic-editing occurs, we need to update the set of known
+    // senders in each stream/topic pair.  This is complicated by the
+    // fact that the event we receive from the server does not
+    // communicate which senders were present before-and-after; so our
+    // strategy is to just rebuild the data structure for the topic
+    // from message_store data.
+
+    const topic_dict = topic_senders.get(stream_id);
+
+    if (new_stream_id !== undefined) {
+        const old_stream_senders = topic_dict.get(old_topic);
+        const new_stream_topic_dict = topic_senders.get(new_stream_id) || new Map();
+        const new_stream_senders = new_stream_topic_dict.get(new_topic);
+        if (new_stream_senders === undefined) {
+            // We only moved the entire topic to a new stream,
+            // there were no change in senders.
+            new_stream_topic_dict.set(new_topic, old_stream_senders);
+            topic_senders.set(new_stream_id, new_stream_topic_dict);
+            topic_dict.delete(old_topic);
+            topic_senders.set(stream_id, topic_dict);
+            return;
+        }
+    }
```

Comment:
I don't understand why you re-adding this block.  I think I failed to post a comment that explained my changes when I pushed to this PR, but I believe I arranged the code below such that it'll do the right thing without this special/duplicate block.

Fundamentally, the thing we're doing is the same is all cases -- and rebuilding the state from the old stream/topic pair and doing the same for the new stream/topic pair, and it doesn't matter whether they're on the same stream or not.

For discussion, this is the diff from what I'd pushed:
```
-exports.process_topic_edit = function (old_stream_id, old_topic, new_topic, new_stream_id) {
+exports.process_topic_edit = function (stream_id, old_topic, new_topic, new_stream_id) {
     // When topic-editing occurs, we need to update the set of known
     // senders in each stream/topic pair.  This is complicated by the
     // fact that the event we receive from the server does not
@@ -40,9 +40,26 @@ exports.process_topic_edit = function (old_stream_id, old_topic, new_topic, new_
     // strategy is to just rebuild the data structure for the topic
     // from message_store data.
 
+    const topic_dict = topic_senders.get(stream_id);
+
+    if (new_stream_id !== undefined) {
+        const old_stream_senders = topic_dict.get(old_topic);
+        const new_stream_topic_dict = topic_senders.get(new_stream_id) || new Map();
+        const new_stream_senders = new_stream_topic_dict.get(new_topic);
+        if (new_stream_senders === undefined) {
+            // We only moved the entire topic to a new stream,
+            // there were no change in senders.
+            new_stream_topic_dict.set(new_topic, old_stream_senders);
+            topic_senders.set(new_stream_id, new_stream_topic_dict);
+            topic_dict.delete(old_topic);
+            topic_senders.set(stream_id, topic_dict);
+            return;
+        }
+    }
+
     // This removes the old topic_dict
-    const old_topic_dict = topic_senders.get(old_stream_id);
-    old_topic_dict.delete(old_topic);
+    topic_dict.delete(old_topic);
+    topic_senders.set(stream_id, topic_dict);
 
     // Re-processing every message in both the old and new topics is
     // expensive.  It also potentially loses data, because
@@ -51,12 +68,15 @@ exports.process_topic_edit = function (old_stream_id, old_topic, new_topic, new_
     // `message_store` from the server (E.g. from when we narrowed to
     // a stream).  But it's the most correct implementation we can
     // sensibly do with existing data structures.
-    const old_topic_msgs = message_util.get_messages_in_topic(old_stream_id, old_topic);
+    const old_topic_msgs = message_util.get_messages_in_topic(stream_id, old_topic);
     old_topic_msgs.forEach((msg) => {
         exports.process_message_for_senders(msg);
     });
 
-    const new_topic_msgs = message_util.get_messages_in_topic(new_stream_id, new_topic);
+    // if topic's stream_id was changed, use the new stream_id
+    stream_id = new_stream_id || stream_id;
+
+    const new_topic_msgs = message_util.get_messages_in_topic(stream_id, new_topic);
     new_topic_msgs.forEach((msg) => {
         exports.process_message_for_senders(msg);
     });
```

Reply:
Yeah! I feel silly. :+1: